module/zfs/dnode.c

```diff
@@ -1646,20 +1646,29 @@ dnode_try_claim(objset_t *os, uint64_t object, int slots)
 }
 
 /*
- * Checks if the dnode contains any uncommitted dirty records.
+ * Checks if the dnode might contain any uncommitted changes to data blocks.
+ * Dirty metadata (e.g. bonus buffer) does not count.
```

Comment:
I believe one exception is the file size.  For example, if you `truncate(2)` or `ftruncate(2)` a new file to a non-zero length `lseek(2)` won't report the hole.  This appears to be what's happening with the failing `mmap_seek_001_pos` test case.  I didn't verify this, but smaller is probably fine since it should create frees to process.